File: DSA/algorithms/count_construct.py
# Write a function canConstruct(target, wordBank)
# The function should return a boolean indicating whether or not the target can be constructed by concatenating elements of the wordBank array
# You may reuse elements of the wordBank as many times as necessary

def count_construct(target, word_bank, memo=None):
    if memo == None:
        memo = dict()
    if target in memo:
        return memo[target]
    if target == '':
        return 1

    total_ways = 0

    for word in word_bank:
        if target.startswith(word):
            total_ways += count_construct(target[len(word):], word_bank, memo)

    memo[target] = total_ways
    return total_ways


if __name__ == '__main__':
    print(count_construct('purple', ['purp', 'p', 'ur', 'le', 'purpl']))
    print(count_construct('abcdef', ['ab', 'abc', 'cd', 'def', 'abcd']))
    print(count_construct('skateboard', ['bo', 'rd', 'ate', 't', 'ska', 'sk', 'boar']))
    print(count_construct('enterapotentpot', ['a', 'p', 'ent', 'enter', 'ot','o', 't']))
    print(count_construct('eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef', ['e', 'ee', 'eee', 'eeee', 'eeeee', 'eeeeee']))

File: DSA/algorithms/countConstruct.js
// Write a function countConstruct(target, wordBank)
// The function should return the number of ways that the target can be constructed by concatenating elements of the wordbank array
// You may resuse elements of the wordbank as many times as necessary

function countConstruct(target, wordBank, memo={}) {
	if (target in memo) return memo[target]
	if (target==='') return 1
	
	let totalCount = 0

	for (let word of wordBank) {
		if (target.indexOf(word)===0) {
			totalCount += countConstruct(target.slice(word.length), wordBank, memo)
		}
	}

	memo[target] = totalCount
	return totalCount
}

console.log(countConstruct('purple', ['purp', 'p', 'ur', 'le', 'purpl']))
console.log(countConstruct('abcdef', ['ab', 'abc', 'cd', 'def', 'abcd']))
console.log(countConstruct('skateboard', ['bo', 'rd', 'ate', 't', 'ska', 'sk', 'boar']))
console.log(countConstruct('enterapotentpot', ['a', 'p', 'ent', 'enter', 'ot', 'o', 't']))
console.log(countConstruct('eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeef', ['e', 'ee', 'eee', 'eeee', 'eeeee', 'eeeeee']))
